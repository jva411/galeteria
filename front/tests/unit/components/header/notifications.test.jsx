/* eslint-disable no-undef */
import React from 'react';
import { DateTime } from 'luxon';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Notifications from '../../../../components/header/notifications';

describe('unit tests of component \'notification\'', () => {
    const mockProps = {
        notifications: {
            datasets: {
                data: [
                    {
                        message: 'Dataset not found.',
                        name: '70d4b5d8-080d-4bd5-b05a-ed83f7943b61',
                        operation: 'GET',
                        /* eslint-disable camelcase */
                        request_time: DateTime.fromMillis(Date.parse('2020-10-21 16:14:43.981000-03:00')),
                        status: 'WARNING'
                    }
                ],
                seen: true
            },
            models: {
                data: [
                    {
                        message: 'Model not found.',
                        name: 'ce06865f-6673-4b8d-b423-2a2e7faa01b6',
                        operation: 'GET',
                        /* eslint-disable camelcase */
                        request_time: DateTime.fromMillis(Date.parse('2020-10-21 16:14:45.457000-03:00')),
                        status: 'INFO'
                    }
                ],
                seen: true
            },
            users: {
                data: [
                    {
                        message: 'User can not be deleted. Try again later.',
                        name: 'BalEjpTdgOyXjqJiqcfg',
                        operation: 'DELETE',
                        /* eslint-disable camelcase */
                        request_time: DateTime.fromMillis(Date.parse('2020-10-21 16:14:42.537000-03:00')),
                        status: 'ERROR'
                    }
                ],
                seen: true
            },
            groups: {
                data: [
                    {
                        message: 'Group can not be deleted. Try again later.',
                        name: 'GroupName',
                        operation: 'DELETE',
                        /* eslint-disable camelcase */
                        request_time: DateTime.fromMillis(Date.parse('2020-10-20 16:11:35.254000-03:00')),
                        status: 'SUCCESS'
                    }
                ],
                seen: true
            }
        }
    }

    test('should render notifications button', () => {
        render(<Notifications />)

        const button = screen.getByRole('button')

        expect(button).toBeInTheDocument()
        expect(button).toHaveAttribute('title', 'notifications')
    })

    test('should hidde tabs when button is not clicked', () => {
        render(<Notifications />)

        expect(screen.getByTestId('popover-content').style).toHaveProperty('visibility', 'hidden')
    })

    test('should show tabs when button is clicked', () => {
        render(<Notifications />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.getByTestId('popover-content').style).toHaveProperty('visibility', 'visible')
    })

    test('should not render the tabs when popover is not open', () => {
        render(<Notifications />)

        expect(screen.queryByText(/models/iu)).toBeNull()
        expect(screen.queryByText(/datasets/iu)).toBeNull()
        expect(screen.queryByText(/users/iu)).toBeNull()
        expect(screen.queryByText(/groups/iu)).toBeNull()
    })

    test('should render 4 tabs when popover is open', () => {
        render(<Notifications />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.getByText(/models/iu)).toBeInTheDocument()
        expect(screen.getByText(/datasets/iu)).toBeInTheDocument()
        expect(screen.getByText(/users/iu)).toBeInTheDocument()
        expect(screen.getByText(/groups/iu)).toBeInTheDocument()
    })

    test('should render \'No notifications\' when not have notifications', () => {
        render(<Notifications />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.getByText(/no notifications/iu)).toBeInTheDocument()
    })

    test('should not render \'No notifications\' when have notifications', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.queryByText(/no notifications/iu)).toBeNull()
    })

    test('should render \'Clear all notifications\' when have notifications', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.getByText(/clear all notifications/iu)).toBeInTheDocument()
    })

    test('should not render \'Clear all notifications\' when not have notifications', () => {
        render(<Notifications />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.queryByText(/clear all notifications/iu)).toBeNull()
    })

    test('should hidde red highlight when all notications have been seen', () => {
        render(<Notifications {...mockProps} />)

        expect(screen.getByTestId('notification-highlight')).not.toBeVisible()
    })

    test('should show red highlight when there is a notication that has not been seen', () => {
        mockProps.notifications.datasets.seen = false
        render(<Notifications {...mockProps} />)

        expect(screen.getByTestId('notification-highlight')).toBeVisible()
        mockProps.notifications.datasets.seen = true
    })

    test('should render models notifications when models tab is selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.getByText(/Model not found/iu)).toBeInTheDocument()
    })

    test('should not render models notifications when models tab is not selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))
        userEvent.click(screen.getByRole('tab', { name: /datasets/iu }))

        expect(screen.queryByText(/Model not found/iu)).toBeNull()
    })

    test('should render datasets notifications when datasets tab is selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))
        userEvent.click(screen.getByRole('tab', { name: /datasets/iu }))

        expect(screen.getByText(/Dataset not found/iu)).toBeInTheDocument()
    })

    test('should not render datasets notifications when datasets tab is not selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.queryByText(/Dataset not found/iu)).toBeNull()
    })

    test('should render users notifications when users tab is selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))
        userEvent.click(screen.getByRole('tab', { name: /users/iu }))

        expect(screen.getByText(/user can not be deleted/iu)).toBeInTheDocument()
    })

    test('should not render users notifications when users tab is not selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.queryByText(/user can not be deleted/iu)).toBeNull()
    })

    test('should render groups notifications when groups tab is selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))
        userEvent.click(screen.getByRole('tab', { name: /groups/iu }))

        expect(screen.getByText(/group can not be deleted/iu)).toBeInTheDocument()
    })

    test('should not render groups notifications when groups tab is not selected', () => {
        render(<Notifications {...mockProps} />)

        userEvent.click(screen.getByRole('button'))

        expect(screen.queryByText(/group can not be deleted/iu)).toBeNull()
    })

    test('should Notifications match with Snapshot', () => {
        const tree = renderer.create(<Notifications />).toJSON()

        expect(tree).toMatchSnapshot()
    })
})
