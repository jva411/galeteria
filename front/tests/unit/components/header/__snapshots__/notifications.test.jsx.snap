// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit tests of component 'notification' should Notifications match with Snapshot 1`] = `
Array [
  <div
    data-testid="notification-highlight"
    hidden={true}
  />,
  <button
    aria-controls="popover-content-144"
    aria-expanded={false}
    aria-haspopup="dialog"
    className="chakra-button css-zse23x"
    id="popover-trigger-144"
    onClick={[Function]}
    title="notifications"
    type="button"
  >
    <svg
      aria-hidden={true}
      fill="currentColor"
      focusable={false}
      height="2.4rem"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="2.4rem"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 13.586V10c0-3.217-2.185-5.927-5.145-6.742C13.562 2.52 12.846 2 12 2s-1.562.52-1.855 1.258C7.185 4.074 5 6.783 5 10v3.586l-1.707 1.707C3.105 15.48 3 15.734 3 16v2c0 .553.447 1 1 1h16c.553 0 1-.447 1-1v-2c0-.266-.105-.52-.293-.707L19 13.586zM19 17H5v-.586l1.707-1.707C6.895 14.52 7 14.266 7 14v-4c0-2.757 2.243-5 5-5s5 2.243 5 5v4c0 .266.105.52.293.707L19 16.414V17zM12 22c1.311 0 2.407-.834 2.818-2H9.182C9.593 21.166 10.689 22 12 22z"
      />
    </svg>
  </button>,
  <div
    className="chakra-popover__popper css-0"
    style={
      Object {
        "--popper-arrow-shadow-color": undefined,
        "--popper-arrow-size": "2rem",
        "minWidth": "max-content",
        "position": "absolute",
        "visibility": "hidden",
      }
    }
  >
    <section
      className="chakra-popover__content css-12f8fva"
      data-testid="popover-content"
      id="popover-content-144"
      onBlur={[Function]}
      onKeyDown={[Function]}
      role="dialog"
      style={
        Object {
          "opacity": 0,
          "transform": "scale(0.95) translateZ(0)",
          "transformOrigin": "var(--popper-transform-origin)",
          "visibility": "hidden",
        }
      }
      tabIndex={-1}
    />
  </div>,
  <hr
    aria-orientation="vertical"
    className="chakra-divider css-2nwp48"
  />,
]
`;
